A54-restrict-control-plane-status-codes.md

```diff
@@ -0,0 +1,136 @@
+broken status code usages.
+
+Now, however, it is more common to implement custom name resolvers and load
+balancers and for those components to use gRPC themselves. This happened first
+with gRPC-LB, but also happens with xDS, Route Lookup Service (RLS), and others.
+It is now harder to audit all the code running in the control plane and it is
+very hard to find status "leaks" from a remote control plane server into the
+data plane. This is a serious risk as control plane errors can't be copied
+verbatim to the data plane; they need to be reinterpreted for the new context
+(generally by becoming `UNAVAILABLE` and attaching additional context).
+
+As an example, if a name resolver is looking up the host `example.com`, that
+could be implemented as a gRPC call to a service. That service might fail the
+RPC with `NOT_FOUND` because the host is not registered. But if the data plane
+has an RPC reading a file `funny-jokes.txt`, the RPC should fail with
+`UNAVAILABLE`. Failing the RPC with `NOT_FOUND` would confuse the application
+into believing `funny-jokes.txt` doesn't exist. This is the situation that
+caused a serious [Spotify outage][] because of [a bug in gRPC Java][issue 8950].
+
+[statuscodes.md]: https://github.com/grpc/grpc/blob/master/doc/statuscodes.md
+[Spotify outage]: https://engineering.atspotify.com/2022/03/incident-report-spotify-outage-on-march-8/
+[issue 8950]: https://github.com/grpc/grpc-java/issues/8950
+
+## Proposal
+
+Add checks to each language's Channel that prevents the control plane from
+passing "inappropriate statuses" to the data plane. Specifically:
+
+ * When applying a failed Service Config parsing result from the name resolver
+ * When applying a failed Config Selector result from the name resolver (an
+   internal API for xDS)
+ * When applying a failed Picker result from the load balancer
+
+The check locations are precise and apply only to the places where the control
+plane passes status codes to the data plane. At the time of this gRFC, this list
+is exhaustive. Future gRFCs should add similar requirements in the case they add
+an API that passes status codes from the control plane to the data plane.
+
+"Inappropriate statuses" are statuses whose code is one of:
+
+ * `OK` (in some APIs this is okay, but it must not be copied to an RPC)
+ * `INVALID_ARGUMENT`
+ * `NOT_FOUND`,
+ * `ALREADY_EXISTS`
+ * `FAILED_PRECONDITION`
+ * `ABORTED`
+ * `OUT_OF_RANGE`
+ * `DATA_LOSS`
+
+If an "inappropriate status" is detected, the status should be rewritten to use
+`INTERNAL` status code and include text within the status message making it
+clear a bug occurred. The original status code and message should be included
+within the new status message so the error information is not lost.
+
+Retry policy status code matching may only be checked after the control plane
+status check and rewrite. However, if an inappropriate status is rewritten and
+the retry policy allows retrying `INTERNAL`, implementations are free to not
```

Comment:
Do you expect any of our implementations to actually do that?  Given the layering involved, it seems like we'd need to do extra work to *not* subject the failure to the normal configured retry policy, and I wasn't expecting that we had any reason to do that.

It might be clearer to just say something like "Note that a pick failure will (or "may", if there's a reason to be flexible) be subject to the configured retry policy, if any, regardless of whether the status code returned by the picker was an allowed code."

Reply:
Our implementations probably won't, but I'm making it clear there are no guarantees here. We had explicitly stated in our discussions that we didn't really care the impact to retries. I need to say that the retry policy needs to be protected from these bugs, but I didn't want to imply that the retry policy must be consulted. There's really no guarantees there (even if all our implementations are likely to behave the same way).

I don't know what you're trying to achieve with the "Note that a pick failure..." language. That's the opposite of what I was trying to accomplish. If I was okay with that, I'd just delete this sentence without replacement.

Reply:
I guess I'm wondering what value we get out of allowing implementations to not retry.  It's not clear to me when we'd actually want to do that in practice, and I suspect the wording here is going to cause confusion over the actual behavior that users should expect.  It's not clear to me that the flexibility is worth that potential confusion.

My preference here would be to just say that the retry policy will be applied to the rewritten `INTERNAL`.

Alternatively, if we do want to allow implementations to not retry, how about replacing "implementations are free to not retry the RPC" with "implementations may retry the RPC"?  The word "may" implies that it's not mandatory, but the wording is positive instead of negative, which seems to more clearly point to the behavior we actually expect our implementations to exhibit.

Reply:
> I suspect the wording here is going to cause confusion over the actual behavior that users should expect

The point of the wording here was that there is no expected behavior.

> implementations may retry the RPC

That has its own confusions. I inverted it to be "may fail the RPC" which worked fine, but I decided to just remove it. The "may only" already allows such behavior (although not obviously so) and it doesn't matter much.